macros/src/main/scala/org/zalando/grafter/macros/ReaderMacro.scala

```diff
@@ -13,43 +12,42 @@ object ReaderMacro {
 
   val annotation = "reader"
 
-  def impl(c: whitebox.Context)(annottees: c.Expr[Any]*): c.Expr[Any] = {
-    import c.universe._
-    def name(t: Name) = t.decodedName.toString.trim
-
-    val (classTree, companionTree): (Tree, Option[Tree]) =
-      annotationInputs(c)(annotation)(annottees)
-
-    val genericTypeName = internal.reificationSupport.freshTypeName("A")
-    val ClassDef(_, className, _, Template(_, _, fields)) = c.typecheck(classTree)
-
-    val params = fieldsNamesAndTypes(c)(fields)
-
-    val implicitParameters = params
-      .map { case (fieldName, fieldType) =>
-        val readerName = TermName(name(fieldName)+"Reader")
-        c.Expr[ValDef](q"""$readerName: cats.data.Reader[$genericTypeName, $fieldType]""")
-      }
-
-    val paramNames = params.map(_._1)
-    val readValues = paramNames.map { p => q"""val ${TermName("_"+name(p)+"Value")} = ${TermName(name(p)+"Reader")}.apply(r);""" }
-    val values     = paramNames.map { p => q"""${TermName("_"+name(p)+"Value")}""" }
-    val klassName  = name(className)
-
-    outputs(c)(classTree, className, companionTree) {
-      q"""
-         implicit def reader[$genericTypeName](implicit ..$implicitParameters): cats.data.Reader[$genericTypeName, $className] = {
-           cats.data.Reader { r =>
-             ..$readValues
-             new ${TypeName(klassName)}(...${List(values)})
-           }
-         }
-       """
+  def expand(classDef: Defn.Class, objectDef: Option[Defn.Object]): Term.Block = {
+    classDef.ctor.paramss.toList match {
+      case Nil =>
+        output(classDef, objectDef)()
+
+      case params :: _ =>
+        val parameters = collectParamTypesAndNamesAsList(params)
+
+        val implicitParameters = parameters.map { case (fieldType, fieldName) =>
+          param"""${Term.Name(fieldName.value+"Reader")}: cats.data.Reader[A, $fieldType]"""
+        }.map { p => p.copy(p.mods :+ Mod.Implicit()) }
+
+        val paramNames = parameters.map(_._2)
+        val readValues = paramNames.map { p => q"""val ${Pat.Var.Term(Term.Name("_"+p.value+"Value"))} = ${Term.Name(p.value+"Reader")}.apply(r);""" }
+        val values     = paramNames.map { p => q"""${Term.Name("_"+p.value+"Value")}""" }
+
+        output(classDef, objectDef) {
+          q"""
+            implicit def reader[A](..$implicitParameters): cats.data.Reader[A, ${classDef.name}] = {
+              cats.data.Reader { r =>
+                ..$readValues
```

Comment:
What does those dots mean? Any specific documentation I should look at to understand this?